File: apps/android/app/src/main/java/chat/mural/network/OpenRouterAPIClient.kt
package chat.mural.network

import android.util.Base64
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import okhttp3.Call
import okhttp3.Callback
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okio.Buffer

/**
 * OpenRouter-backed teaching, STT (Qwen3 ASR Flash) and TTS (Kokoro 82M).
 * Uses the same credential store as BYOK; keys must start with sk-or-.
 */
class OpenRouterAPIClient private constructor(
    private val readCredential: () -> String?,
    private val client: OkHttpClient = defaultClient(),
    private val cache: PhraseAudioCache? = null,
) : TeachingClient {
    constructor(credentials: CredentialStore, cache: PhraseAudioCache) : this(credentials::read, defaultClient(), cache)
    internal constructor(key: String?, client: OkHttpClient, cache: PhraseAudioCache? = null) :
        this({ key }, client, cache)

    override suspend fun respond(
        instructions: String,
        input: String,
        schema: JsonObject?,
        search: Boolean,
        purpose: HelperPurpose?,
    ): APIResult {
        if (search) {
            // Web search tools are not wired for OpenRouter in this build; callers should degrade gracefully.
            throw APIClient.APIException.Refused
        }
        val body = buildJsonObject {
            put("model", OpenRouterModels.LLM)
            put("messages", buildJsonArray {
                add(buildJsonObject {
                    put("role", "system")
                    put("content", instructions)
                })
                add(buildJsonObject {
                    put("role", "user")
                    put("content", input)
                })
            })
            put("max_tokens", if (schema == null) 1_400 else 2_200)
            if (schema != null) {
                put("response_format", buildJsonObject {
                    put("type", "json_schema")
                    put("json_schema", buildJsonObject {
                        put("name", "mural_result")
                        put("strict", true)
                        put("schema", schema)
                    })
                })
            }
        }
        return decodeOpenRouterChat(post("chat/completions", body))
    }

    /** Voice turn: DeepSeek V4 Pro with brevity system overlay and prior transcript messages. */
    suspend fun voiceReply(
        systemInstructions: String,
        overlay: String,
        history: List<Pair<String, String>>,
        locale: String?,
    ): String {
        val body = buildJsonObject {
            put("model", OpenRouterModels.LLM)
            put("max_tokens", 220)
            put("temperature", 0.7)
            put("messages", buildJsonArray {
                add(buildJsonObject {
                    put("role", "system")
                    put("content", buildString {
                        append(systemInstructions)
                        append("\n\n")
                        append(OpenRouterModels.VOICE_BREVITY_PROMPT)
                        if (overlay.isNotBlank()) {
                            append("\n\nAktuelle Anweisung (nicht vorlesen): ")
                            append(overlay.take(1200))
                        }
                    })
                })
                history.takeLast(24).forEach { (role, content) ->
                    add(buildJsonObject {
                        put("role", role)
                        put("content", content.take(4_000))
                    })
                }
            })
        }
        return decodeOpenRouterChat(post("chat/completions", body)).text.trim()
    }

    suspend fun transcribe(pcm16Mono: ByteArray, sampleRate: Int): String {
        val wav = WavEncoder.encodePcm16Mono(pcm16Mono, sampleRate)
        val base64 = Base64.encodeToString(wav, Base64.NO_WRAP)
        val body = buildJsonObject {
            put("model", OpenRouterModels.STT)
            put("data", base64)
        }
        val response = post("audio/transcriptions", body)
        val text = (response["text"] as? JsonPrimitive)?.contentOrNull?.trim()
            ?: (response["transcript"] as? JsonPrimitive)?.contentOrNull?.trim()
        if (text.isNullOrBlank()) throw APIClient.APIException.InvalidResponse
        return text
    }

    suspend fun synthesizeSpeech(text: String, locale: String?): ByteArray {
        val voice = OpenRouterModels.ttsVoiceForLocale(locale)
        cache?.lookup(OpenRouterModels.TTS, voice, text)?.let { return it }
        val body = buildJsonObject {
            put("model", OpenRouterModels.TTS)
            put("input", text.take(900))
            put("voice", voice)
        }
        val bytes = postBytes("audio/speech", body)
        cache?.store(OpenRouterModels.TTS, voice, text, bytes)
        return bytes
    }

    private suspend fun post(path: String, body: JsonObject): JsonObject {
        val payload = postBytes(path, body)
        return try {
            Json.parseToJsonElement(payload.decodeToString()).jsonObject
        } catch (_: Exception) {
            throw APIClient.APIException.InvalidResponse
        }
    }

    private suspend fun postBytes(path: String, body: JsonObject): ByteArray {
        if (!VALID_PATH.matches(path) || path.contains("..") || path.startsWith('/')) {
            throw APIClient.APIException.InvalidResponse
        }
        val key = readCredential() ?: throw APIClient.APIException.MissingKey
        val request = Request.Builder()
            .url(OPENROUTER_BASE.newBuilder().addPathSegments(path).build())
            .header("Authorization", "Bearer $key")
            .header("Content-Type", JSON_MEDIA_TYPE.toString())
            .header("HTTP-Referer", "https://mural.chat/")
            .header("X-OpenRouter-Title", "Mural Android")
            .post(body.toString().toRequestBody(JSON_MEDIA_TYPE))
            .build()
        return suspendCancellableCoroutine { continuation ->
            val call = client.newCall(request)
            continuation.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onFailure(call: Call, error: IOException) {
                    if (continuation.isActive) continuation.resumeWithException(error)
                }
                override fun onResponse(call: Call, response: Response) {
                    try {
                        val value = response.use {
                            if (it.code !in 200..299) throw APIClient.APIException.Http(it.code)
                            it.readBoundedBody()
                        }
                        if (continuation.isActive) continuation.resume(value)
                    } catch (error: Exception) {
                        if (continuation.isActive) continuation.resumeWithException(error)
                    }
                }
            })
        }
    }

    private fun Response.readBoundedBody(): ByteArray {
        val responseBody = body ?: throw APIClient.APIException.InvalidResponse
        if (responseBody.contentLength() > MAX_RESPONSE_BYTES) throw APIClient.APIException.InvalidResponse
        val source = responseBody.source()
        val buffer = Buffer()
        var total = 0L
        while (true) {
            val count = source.read(buffer, minOf(8_192L, MAX_RESPONSE_BYTES + 1L - total))
            if (count == -1L) break
            total += count
            if (total > MAX_RESPONSE_BYTES) throw APIClient.APIException.InvalidResponse
        }
        return buffer.readByteArray()
    }

    companion object {
        private val OPENROUTER_BASE = HttpUrl.Builder()
            .scheme("https")
            .host(OpenRouterModels.BASE_HOST)
            .addPathSegment("api")
            .addPathSegment("v1")
            .addPathSegment("")
            .build()
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
        private val VALID_PATH = Regex("[a-z0-9][a-z0-9_/-]*")
        private const val MAX_RESPONSE_BYTES = 2_000_000L
        private val Json = Json { ignoreUnknownKeys = true }

        private fun defaultClient() = OkHttpClient.Builder()
            .connectTimeout(45, TimeUnit.SECONDS)
            .readTimeout(90, TimeUnit.SECONDS)
            .writeTimeout(45, TimeUnit.SECONDS)
            .callTimeout(120, TimeUnit.SECONDS)
            .followRedirects(false)
            .followSslRedirects(false)
            .cookieJar(CookieJar.NO_COOKIES)
            .cache(null)
            .build()
    }
}

/** Minimal PCM16 mono WAV encoder for ASR upload. */
internal object WavEncoder {
    fun encodePcm16Mono(pcm: ByteArray, sampleRate: Int): ByteArray {
        val channels = 1
        val bits = 16
        val byteRate = sampleRate * channels * bits / 8
        val blockAlign = channels * bits / 8
        val dataSize = pcm.size
        val buffer = java.io.ByteArrayOutputStream(44 + dataSize)
        buffer.write("RIFF".toByteArray())
        buffer.write(intLe(36 + dataSize))
        buffer.write("WAVE".toByteArray())
        buffer.write("fmt ".toByteArray())
        buffer.write(intLe(16))
        buffer.write(shortLe(1))
        buffer.write(shortLe(channels))
        buffer.write(intLe(sampleRate))
        buffer.write(intLe(byteRate))
        buffer.write(shortLe(blockAlign))
        buffer.write(shortLe(bits))
        buffer.write("data".toByteArray())
        buffer.write(intLe(dataSize))
        buffer.write(pcm)
        return buffer.toByteArray()
    }

    private fun intLe(value: Int): ByteArray = byteArrayOf(
        (value and 0xff).toByte(),
        (value shr 8 and 0xff).toByte(),
        (value shr 16 and 0xff).toByte(),
        (value shr 24 and 0xff).toByte(),
    )

    private fun shortLe(value: Int): ByteArray = byteArrayOf(
        (value and 0xff).toByte(),
        (value shr 8 and 0xff).toByte(),
    )
}

File: apps/ios/App/OpenRouterAPIClient.swift
import Foundation
import MuralCore

@MainActor final class OpenRouterAPIClient {
    private let session: URLSession
    init() {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 60
        config.timeoutIntervalForResource = 120
        config.httpCookieStorage = nil
        config.urlCache = nil
        session = URLSession(configuration: config)
    }

    func respond(instructions: String, input: String, schema: [String: Any]? = nil) async throws -> APIResult {
        guard let key = CredentialStore.read(), OpenRouterModels.isOpenRouterKey(key) else { throw APIClient.APIError.missingKey }
        var body: [String: Any] = [
            "model": OpenRouterModels.llm,
            "messages": [
                ["role": "system", "content": instructions],
                ["role": "user", "content": input],
            ],
            "max_tokens": schema == nil ? 1400 : 2200,
        ]
        if let schema {
            body["response_format"] = [
                "type": "json_schema",
                "json_schema": ["name": "mural_result", "strict": true, "schema": schema],
            ]
        }
        let json = try await post(path: "chat/completions", key: key, body: body)
        guard let choices = json["choices"] as? [[String: Any]],
              let message = choices.first?["message"] as? [String: Any],
              let text = message["content"] as? String, !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw APIClient.APIError.invalidResponse
        }
        let usage = json["usage"] as? [String: Any]
        return APIResult(
            text: text,
            sources: [],
            usage: APIUsage(
                input: usage?["prompt_tokens"] as? Int ?? 0,
                output: usage?["completion_tokens"] as? Int ?? 0
            )
        )
    }

    func voiceReply(system: String, overlay: String, history: [(role: String, content: String)]) async throws -> String {
        guard let key = CredentialStore.read() else { throw APIClient.APIError.missingKey }
        var messages: [[String: Any]] = [[
            "role": "system",
            "content": system + "\n\n" + OpenRouterModels.voiceBrevity + (overlay.isEmpty ? "" : "\n\n" + overlay),
        ]]
        for item in history.suffix(24) {
            messages.append(["role": item.role, "content": String(item.content.prefix(4000))])
        }
        let json = try await post(path: "chat/completions", key: key, body: [
            "model": OpenRouterModels.llm,
            "max_tokens": 220,
            "temperature": 0.7,
            "messages": messages,
        ])
        guard let choices = json["choices"] as? [[String: Any]],
              let text = choices.first?["message"] as? [String: Any],
              let content = text["content"] as? String else { throw APIClient.APIError.invalidResponse }
        return content.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func synthesizeSpeech(_ text: String) async throws -> Data {
        guard let key = CredentialStore.read() else { throw APIClient.APIError.missingKey }
        var request = URLRequest(url: OpenRouterModels.baseURL.appendingPathComponent("audio/speech"))
        request.httpMethod = "POST"
        request.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("https://mural.chat/", forHTTPHeaderField: "HTTP-Referer")
        request.setValue("Mural iOS", forHTTPHeaderField: "X-OpenRouter-Title")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "model": OpenRouterModels.tts,
            "input": String(text.prefix(900)),
            "voice": OpenRouterModels.defaultVoice,
        ])
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw APIClient.APIError.http((response as? HTTPURLResponse)?.statusCode ?? 0)
        }
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let b64 = json["audio"] as? String, let decoded = Data(base64Encoded: b64) { return decoded }
        return data
    }

    private func post(path: String, key: String, body: [String: Any]) async throws -> [String: Any] {
        var request = URLRequest(url: OpenRouterModels.baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("https://mural.chat/", forHTTPHeaderField: "HTTP-Referer")
        request.setValue("Mural iOS", forHTTPHeaderField: "X-OpenRouter-Title")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw APIClient.APIError.http((response as? HTTPURLResponse)?.statusCode ?? 0)
        }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw APIClient.APIError.invalidResponse }
        return json
    }
}
